src/Servers/HttpSys/src/RequestProcessing/RequestContext.FeatureCollection.cs

```diff
@@ -98,11 +99,24 @@ private enum Fields
         TraceIdentifier = 0x200,
     }
 
-    protected internal void InitializeFeatures()
+    protected internal bool InitializeFeatures()
     {
         _initialized = true;
 
-        Request = new Request(this);
+        // Get the ID before creating the Request object as the Request ctor releases the native memory
+        // We want the ID in case request processing fails and we need the ID to cancel the native request
+        _requestId = RequestId;
+        try
+        {
+            Request = new Request(this);
+        }
+        catch (Exception ex)
+        {
+            Log.RequestParsingError(Logger, ex);
+            Server.SendError(_requestId.Value, StatusCodes.Status400BadRequest, authChallenges: null);
```

Comment:
It's not immediately obvious that this is a safe place to call `SendError`.  What does it do?  Is it synchronous?  Does it schedule something?  It doesn't depend on anything established by `InitializeFeatures`?

Reply:
It's a synchronous call, only state it needs from InitializeFeature is the request ID which we pass in. Would you like some sort of comment here mentioning it's a synchronous call to native code?